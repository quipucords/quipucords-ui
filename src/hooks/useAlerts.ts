/**
 * This custom React Hook provides functions to manage alerts in your application.
 * It allows you to add and remove alerts with specific titles, variants, and unique keys.
 * @module useAlerts
 */
import React from 'react';
import { AlertProps } from '@patternfly/react-core';

const useAlerts = () => {
  const [alert, setAlerts] = React.useState<Partial<AlertProps>[]>([]);

  /**
   * Add an Alert
   *
   * This function adds an alert to the list of alerts.
   *
   * @param {string} title - The title or content of the alert.
   * @param {AlertProps['variant']} variant - The variant or style of the alert (e.g., 'success', 'danger').
   * @param {React.Key} key - A unique key to identify the alert.
   */
  const addAlert = (title: string, variant: AlertProps['variant'], key: React.Key) => {
    setAlerts(prevAlerts => [...prevAlerts, { title, variant, key }]);
  };

  /**
   * Remove an Alert by Key
   *
   * This function removes an alert from the list of alerts based on its unique key.
   *
   * @param {React.Key} key - The unique key of the alert to be removed.
   */
  const removeAlert = (key: React.Key) => {
    setAlerts(prevAlerts => [...prevAlerts.filter(alert => alert.key !== key)]);
  };

  return {
    removeAlert,
    addAlert,
    alerts: alert
  };
};

export default useAlerts;
