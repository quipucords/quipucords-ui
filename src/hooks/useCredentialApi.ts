/**
 * Credential API Hook
 *
 * This hook provides functions for for interacting with credential-related API calls,
 * including adding,editing and deleting credentials and managing selected credentials.
 * @module useCredentialApi
 */
import React from 'react';
import axios from 'axios';
import { CredentialType } from '../types/types';

const useCredentialApi = () => {
  const [pendingDeleteCredential, setPendingDeleteCredential] = React.useState<CredentialType>();

  /**
   * Executes a DELETE request to delete a credential.
   *
   * @param {CredentialType} [credential] - (Optional) The credential to be deleted. If not provided, it uses `pendingDeleteCredential`.
   * @returns {AxiosResponse} - The Axios response object representing the result of the request.
   */
  const deleteCredential = (credential?: CredentialType) => {
    return axios.delete(
      `${process.env.REACT_APP_CREDENTIALS_SERVICE}${(credential || pendingDeleteCredential)?.id}/`
    );
  };

  /**
   * Deletes several selected credentials based on user interaction.
   * Add your logic to handle the deletion of selected items within this function.
   */
  const onDeleteSelectedCredentials = () => {
    const selectedItems = [];
    console.log('Deleting selected credentials:', selectedItems);
  };

  return {
    deleteCredential,
    onDeleteSelectedCredentials,
    pendingDeleteCredential,
    setPendingDeleteCredential
  };
};

export default useCredentialApi;
