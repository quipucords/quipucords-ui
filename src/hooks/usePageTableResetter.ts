import * as React from 'react';

const usePageTableResetter = tableState => {
  const lastFilterAndSort = React.useRef<{
    filterValues: typeof tableState.filter.filterValues;
    activeSort: typeof tableState.sort.activeSort;
  }>({
    filterValues: tableState.filter.filterValues,
    activeSort: tableState.sort.activeSort
  });

  React.useEffect(() => {
    const { filterValues } = tableState.filter;
    const { activeSort } = tableState.sort;
    const { pageNumber, setPageNumber } = tableState.pagination;
    const { filterValues: lastFilterValues, activeSort: lastActiveSort } = lastFilterAndSort.current;

    const filterChanged = JSON.stringify(lastFilterValues) !== JSON.stringify(filterValues);
    const sortChanged = JSON.stringify(lastActiveSort) !== JSON.stringify(activeSort);

    if (pageNumber > 1 && (filterChanged || sortChanged)) {
      setPageNumber(1);
    }

    lastFilterAndSort.current = { filterValues, activeSort };
  }, [
    tableState.filter.filterValues,
    tableState.sort.activeSort,
    tableState.pagination.pageNumber,
    tableState.pagination.setPageNumber
  ]);
};

export { usePageTableResetter as default, usePageTableResetter };
