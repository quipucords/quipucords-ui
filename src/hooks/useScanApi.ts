import { useCallback, useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { type AlertProps } from '@patternfly/react-core';
import axios, { AxiosError, type AxiosResponse, isAxiosError, type AxiosRequestConfig } from 'axios';
import helpers from '../helpers';
import apiHelpers from '../helpers/apiHelpers';
import {
  type Scan,
  type ScanJobType,
  type ScanJobsResponse,
  type SourceType,
  type Connections,
  type ReportsAggregateResponse,
  simpleScanJob
} from '../types/types';

type ApiDeleteScanSuccessType = {
  message: string;
  deleted?: number[];
  missing?: number[];
};

type ApiScanErrorType = {
  detail?: string;
  message: string;
};

interface InProgressState {
  state: 'InProgress';
}

interface SuccessfulState {
  state: 'Successful';
  mergedReportId: number;
}

interface ErroredState {
  state: 'Errored';
  errorMessage: string;
}

type MergeProcessState = InProgressState | SuccessfulState | ErroredState;

const useCreateScanApi = (
  onAddAlert: (alert: AlertProps) => void,
  onFieldErrors?: (errors: { [key: string]: string }) => void
) => {
  const { t } = useTranslation();

  const apiCall = useCallback(
    (payload: Scan): Promise<AxiosResponse<Scan>> => axios.post(`${process.env.REACT_APP_SCANS_SERVICE}`, payload),
    []
  );

  const callbackSuccess = useCallback(
    (response: AxiosResponse<Scan>) => {
      onAddAlert({
        title: t('toast-notifications.description', {
          context: 'scan-report_created',
          name: response?.data?.name
        }),
        variant: 'success'
      });
      return response.data;
    },
    [onAddAlert, t]
  );

  const callbackError = useCallback(
    (error: AxiosError<ApiScanErrorType>, name: string) => {
      if (onFieldErrors && apiHelpers.hasFieldValidationErrors(error)) {
        const fieldErrors = apiHelpers.parseFieldErrors(error.response?.data);
        onFieldErrors(fieldErrors);
      } else {
        onAddAlert({
          title: t('toast-notifications.description', {
            context: 'scan-report_created_error',
            name: name,
            message: apiHelpers.extractErrorMessage(error?.response?.data)
          }),
          variant: 'danger'
        });
      }
      return Promise.reject(error);
    },
    [onAddAlert, onFieldErrors, t]
  );

  const createScans = useCallback(
    async (payload: Scan) => {
      let response;
      try {
        response = await apiCall(payload);
      } catch (error) {
        if (isAxiosError(error)) {
          return callbackError(error, payload.name);
        }
        if (!helpers.TEST_MODE) {
          console.error(error);
        }
      }
      return callbackSuccess(response);
    },
    [apiCall, callbackSuccess, callbackError]
  );

  return {
    apiCall,
    callbackError,
    callbackSuccess,
    createScans
  };
};

const useRunScanApi = (
  onAddAlert: (alert: AlertProps) => void,
  onFieldErrors?: (errors: { [key: string]: string }) => void,
  useCreateScan: typeof useCreateScanApi = useCreateScanApi
) => {
  const { t } = useTranslation();
  const { createScans } = useCreateScan(onAddAlert, onFieldErrors);

  const apiCall = useCallback(
    (scanId: string): Promise<AxiosResponse<ScanJobType>> =>
      axios.post(`${process.env.REACT_APP_SCANS_SERVICE}${scanId}/jobs/`),
    []
  );

  const callbackSuccess = useCallback(
    (name: string) => {
      onAddAlert({
        title: t('toast-notifications.description', {
          context: 'scan-report_play',
          name: name
        }),
        variant: 'success'
      });
      return;
    },
    [onAddAlert, t]
  );

  const callbackError = useCallback(
    (error: AxiosError<ApiScanErrorType>, name: string) => {
      if (onFieldErrors && apiHelpers.hasFieldValidationErrors(error)) {
        const fieldErrors = apiHelpers.parseFieldErrors(error.response?.data);
        onFieldErrors(fieldErrors);
      } else {
        onAddAlert({
          title: t('toast-notifications.description', {
            context: 'scan-report_play_error',
            name: name,
            message: apiHelpers.extractErrorMessage(error?.response?.data)
          }),
          variant: 'danger'
        });
      }
      return Promise.reject(error);
    },
    [onAddAlert, onFieldErrors, t]
  );

  const runScans = useCallback(
    async (payload: Scan, isRescan = false) => {
      try {
        const scanId =
          (isRescan && payload?.id?.toString()) ||
          (await createScans(payload)
            .then(result => result?.id?.toString())
            .catch(() => {
              const customError = new Error('Failed to create scan');
              throw customError;
            }));

        if (!scanId) {
          const customError = new Error('No ID returned');
          Object.assign(customError, {
            isAxiosError: true,
            response: {
              data: {
                message: 'No ID returned',
                detail: 'The scan or rescan operation did not return an ID.'
              }
            }
          });
          throw customError;
        }
        await apiCall(scanId);
      } catch (error) {
        if (isAxiosError(error)) {
          return callbackError(error, payload.name);
        }
        if (!helpers.TEST_MODE) {
          console.error(error);
        }
        return Promise.reject(error);
      }
      return callbackSuccess(payload.name);
    },
    [apiCall, callbackSuccess, callbackError, createScans]
  );

  return {
    apiCall,
    callbackError,
    callbackSuccess,
    runScans
  };
};

const useDeleteScanApi = (onAddAlert: (alert: AlertProps) => void) => {
  const { t } = useTranslation();

  const apiCall = useCallback(
    (ids: Scan['id'][]): Promise<AxiosResponse<ApiDeleteScanSuccessType>> =>
      axios.post(`${process.env.REACT_APP_SCANS_SERVICE_BULK_DELETE}`, { ids }),
    []
  );

  const callbackSuccess = useCallback(
    (response: AxiosResponse<ApiDeleteScanSuccessType>, updatedScans: Scan[]) => {
      const { data } = response;

      const missingNames = data?.missing
        ?.map(id => updatedScans.find(scan => id === scan.id)?.name)
        .filter(Boolean)
        .join(', ');

      const missingScansMsg = `The following scans could not be found: ${missingNames}`;

      onAddAlert({
        title: t('toast-notifications.description', {
          context: 'deleted-scan',
          count: updatedScans.length,
          name: updatedScans.map(({ name }) => name).join(', ')
        }),
        variant: 'success'
      });

      if (data?.missing?.length) {
        console.log(missingScansMsg);
      }
      return;
    },
    [onAddAlert, t]
  );

  const callbackError = useCallback(
    ({ message, response }: AxiosError<ApiScanErrorType>, updatedScans: Scan[]) => {
      onAddAlert({
        title: t('toast-notifications.description', {
          context: 'deleted-scan_error',
          count: updatedScans.length,
          name: updatedScans.map(({ name }) => name).join(', '),
          message: response?.data?.detail || response?.data?.message || message || 'Unknown error'
        }),
        variant: 'danger'
      });
      return;
    },
    [onAddAlert, t]
  );

  const deleteScans = useCallback(
    async (scan: Scan | Scan[]) => {
      let response;
      const updatedScans = (Array.isArray(scan) && scan) || [scan];
      try {
        response = await apiCall(updatedScans.map(({ id }) => id));
      } catch (error) {
        if (isAxiosError(error)) {
          return callbackError(error, updatedScans);
        }
        if (!helpers.TEST_MODE) {
          console.error(error);
        }
      }
      return callbackSuccess(response, updatedScans);
    },
    [apiCall, callbackSuccess, callbackError]
  );

  return {
    apiCall,
    callbackError,
    callbackSuccess,
    deleteScans
  };
};

const useGetScanJobsApi = (onAddAlert: (alert: AlertProps) => void) => {
  const { t } = useTranslation();

  const apiCall = useCallback(
    (scanId: number): Promise<AxiosResponse<ScanJobsResponse>> =>
      axios.get(`${process.env.REACT_APP_SCANS_SERVICE}${scanId}/jobs/`, { params: { page_size: 1000 } }),
    []
  );

  const callbackSuccess = useCallback((response: AxiosResponse<ScanJobsResponse>) => {
    return response;
  }, []);

  const callbackError = useCallback(
    (error: AxiosError<ApiScanErrorType>, scanId: number) => {
      onAddAlert({
        title: t('toast-notifications.description', {
          context: 'scan-jobs_fetch_error',
          id: scanId,
          message: apiHelpers.extractErrorMessage(error?.response?.data)
        }),
        variant: 'danger'
      });
      return;
    },
    [onAddAlert, t]
  );

  const getScanJobs = useCallback(
    async (scanId: number) => {
      let response;
      try {
        response = await apiCall(scanId);
      } catch (error) {
        if (isAxiosError(error)) {
          return callbackError(error, scanId);
        }
        if (!helpers.TEST_MODE) {
          console.error(error);
        }
      }
      return callbackSuccess(response);
    },
    [apiCall, callbackSuccess, callbackError]
  );

  return {
    apiCall,
    callbackError,
    callbackSuccess,
    getScanJobs
  };
};

const useGetAggregateReportApi = (onAddAlert: (alert: AlertProps) => void) => {
  const { t } = useTranslation();

  const apiCall = useCallback(
    (reportId: number): Promise<AxiosResponse<ReportsAggregateResponse>> =>
      axios.get(`${process.env.REACT_APP_REPORTS_SERVICE}${reportId}/aggregate/`),
    []
  );

  const callbackSuccess = useCallback(
    (response: AxiosResponse<ReportsAggregateResponse>, reportId: number) => ({
      id: reportId,
      report: response.data
    }),
    []
  );

  const callbackError = useCallback(
    (error: AxiosError<ApiScanErrorType>, reportId: number) => {
      onAddAlert({
        title: t('toast-notifications.description', {
          context: 'report_aggregate_error',
          name: reportId,
          message: apiHelpers.extractErrorMessage(error?.response?.data)
        }),
        variant: 'danger'
      });

      return Promise.reject(error);
    },
    [onAddAlert, t]
  );

  const getAggregateReport = useCallback(
    async (reportId: number) => {
      let response;
      try {
        response = await apiCall(reportId);
      } catch (error) {
        if (isAxiosError(error)) {
          return callbackError(error, reportId);
        }
        if (!helpers.TEST_MODE) {
          console.error(error);
        }
      }
      return callbackSuccess(response, reportId);
    },
    [apiCall, callbackSuccess, callbackError]
  );

  return {
    apiCall,
    callbackSuccess,
    callbackError,
    getAggregateReport
  };
};

const useDownloadReportApi = (onAddAlert: (alert: AlertProps) => void) => {
  const { t } = useTranslation();

  const apiCall = useCallback(
    (reportId: number, { responseType = 'blob' }: AxiosRequestConfig = {}): Promise<AxiosResponse<Blob>> =>
      axios.get(`${process.env.REACT_APP_REPORTS_SERVICE}${reportId}/`, { responseType }),
    []
  );

  const callbackSuccess = useCallback(
    (response: AxiosResponse<Blob>, reportId: number) => {
      onAddAlert({
        title: t('toast-notifications.description', {
          context: 'report_downloaded',
          name: reportId
        }),
        variant: 'success'
      });

      helpers.downloadData(
        response.data,
        `report_id_${reportId}_${new Date()
          .toISOString()
          .replace('T', '_')
          .replace(/[^\d_]/g, '')}.tar.gz`,
        'application/gzip'
      );
      return response;
    },
    [onAddAlert, t]
  );

  const callbackError = useCallback(
    (error: AxiosError<ApiScanErrorType>, reportId: number) => {
      onAddAlert({
        title: t('toast-notifications.description', {
          context: 'report_download_error',
          name: reportId,
          message: apiHelpers.extractErrorMessage(error?.response?.data)
        }),
        variant: 'danger'
      });
      return;
    },
    [onAddAlert, t]
  );

  const downloadReport = useCallback(
    async (reportId: number) => {
      let response;
      try {
        response = await apiCall(reportId);
      } catch (error) {
        if (isAxiosError(error)) {
          return callbackError(error, reportId);
        }
        if (!helpers.TEST_MODE) {
          console.error(error);
        }
      }
      return callbackSuccess(response, reportId);
    },
    [apiCall, callbackSuccess, callbackError]
  );

  return {
    apiCall,
    callbackSuccess,
    callbackError,
    downloadReport
  };
};

const useShowConnectionsApi = () => {
  const apiCall = useCallback((source: SourceType): Promise<AxiosResponse> => {
    return axios.get(`${process.env.REACT_APP_SCAN_JOBS_SERVICE}${source.connection.id}/connection/`, {
      params: { page: 1, page_size: 1000, ordering: 'name', source_type: source.id }
    });
  }, []);

  const callbackSuccess = useCallback(
    (response: AxiosResponse): Connections => ({
      successful: response?.data?.results?.filter((c: { status: string }) => c.status === 'success') || [],
      failed: response?.data?.results?.filter((c: { status: string }) => c.status === 'failed') || [],
      unreachable:
        response?.data?.results?.filter((c: { status: string }) => !['success', 'failed'].includes(c.status)) || []
    }),
    []
  );

  const callbackError = useCallback((error: AxiosError) => Promise.reject(error), []);

  const showConnections = useCallback(
    async (source: SourceType) => {
      let response;
      try {
        response = await apiCall(source);
      } catch (error) {
        if (axios.isAxiosError(error)) {
          return callbackError(error);
        }
        if (!helpers.TEST_MODE) {
          console.error(error);
        }
      }
      return callbackSuccess(response);
    },
    [apiCall, callbackSuccess, callbackError]
  );

  return {
    apiCall,
    callbackError,
    callbackSuccess,
    showConnections
  };
};

const useMergeReportsApi = () => {
  const [mergeProcessState, setMergeProcessState] = useState<MergeProcessState>({ state: 'InProgress' });
  const { t } = useTranslation();
  const [mergeJobId, setMergeJobId] = useState<number | undefined>(undefined);
  const [mergePollAttempt, setMergePollAttempt] = useState<number>(0);

  const waitInterval = process.env.REACT_APP_MERGE_POLL_INTERVAL
    ? Number.parseInt(process.env.REACT_APP_MERGE_POLL_INTERVAL, 10)
    : 1000;

  const requestReportsMerge = useCallback(
    async (report_ids: number[]) => {
      let response: AxiosResponse;
      try {
        response = await axios.post(`${process.env.REACT_APP_REPORTS_SERVICE_MERGE}`, { reports: report_ids });
        const jobId = response.data.job_id;
        if (!jobId) {
          throw new Error(t('merge.error', { context: 'no-jobid' }));
        }
        setMergeJobId(jobId);
      } catch (error) {
        if (isAxiosError(error)) {
          const errorMessage = [error.message, apiHelpers.extractErrorMessage(error.response?.data)].join(': ');
          setMergeProcessState({ state: 'Errored', errorMessage: errorMessage });
        } else if (error instanceof Error) {
          setMergeProcessState({ state: 'Errored', errorMessage: error.message });
        } else {
          setMergeProcessState({ state: 'Errored', errorMessage: t('merge.error', { context: 'unknown' }) });
        }
      }
    },
    [t]
  );

  const cancelReportsMerge = useCallback((): void => {
    setMergeProcessState({ state: 'InProgress' });
    setMergeJobId(undefined);
    setMergePollAttempt(0);
  }, []);

  useEffect(() => {
    if (mergeJobId === undefined) {
      return () => {};
    }

    let pollingTimer: NodeJS.Timeout;

    axios
      .get(`${process.env.REACT_APP_SCAN_JOBS_V2_SERVICE}${mergeJobId}/`)
      .then((response: AxiosResponse<simpleScanJob>) => {
        const status = response.data.status;

        if (status === 'completed') {
          setMergeProcessState({ state: 'Successful', mergedReportId: response.data.report_id });
          return;
        }

        if (['failed', 'canceled'].includes(status)) {
          setMergeProcessState({ state: 'Errored', errorMessage: response.data.status_message });
          return;
        }

        pollingTimer = setTimeout(() => setMergePollAttempt(mergePollAttempt + 1), waitInterval);
      })
      .catch(error => {
        if (isAxiosError(error)) {
          const errorMessage = [error.message, apiHelpers.extractErrorMessage(error.response?.data)].join(': ');
          setMergeProcessState({ state: 'Errored', errorMessage: errorMessage });
        } else {
          setMergeProcessState({ state: 'Errored', errorMessage: t('merge.error', { context: 'unknown' }) });
        }
      });

    return () => {
      clearTimeout(pollingTimer);
    };
  }, [mergeJobId, mergePollAttempt, waitInterval, t]);

  return {
    requestReportsMerge,
    cancelReportsMerge,
    mergeProcessState
  };
};

export {
  useCreateScanApi,
  useRunScanApi,
  useDeleteScanApi,
  useGetAggregateReportApi,
  useGetScanJobsApi,
  useDownloadReportApi,
  useShowConnectionsApi,
  useMergeReportsApi,
  MergeProcessState
};
