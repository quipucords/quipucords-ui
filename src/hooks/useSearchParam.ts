/**
 * A custom hook for managing URL search parameters with React Router, allowing for getting and setting values with ease.
 * Supports default values, optional replacement of the history stack, and automatic unsetting of parameters when set to their default value.
 * @module useSearchParam
 */
import React from 'react';
import { useSearchParams } from 'react-router-dom';

const useSearchParam = (
  name: string,
  defaultValue: string | null = null,
  options?: {
    replace?: boolean;
    unsetWhenDefaultValue?: boolean;
  }
  // eslint-disable-next-line no-unused-vars
): [string | null, (newValue: string) => void, () => void] => {
  const defaultValueRef = React.useRef(defaultValue);
  const [searchParams, setSearchParams] = useSearchParams();
  const value = searchParams.has(name) ? searchParams.get(name) : defaultValueRef.current;
  const unsetWhenDefaultValue = options?.unsetWhenDefaultValue ?? true;
  const replace = options?.replace ?? true;
  const set = React.useCallback(
    (newValue: string) => {
      const newSearchParams = new URLSearchParams(window.location.search);
      if (newSearchParams.get(name) !== newValue) {
        if (unsetWhenDefaultValue && newValue === defaultValueRef.current) {
          newSearchParams.delete(name);
        } else {
          newSearchParams.set(name, newValue);
        }
        setSearchParams(newSearchParams, { replace });
      }
    },
    [name, setSearchParams, unsetWhenDefaultValue, replace]
  );

  const unset = React.useCallback(() => {
    const newSearchParams = new URLSearchParams(window.location.search);
    if (newSearchParams.has(name)) {
      newSearchParams.delete(name);
      setSearchParams(newSearchParams);
    }
  }, [name, setSearchParams]);

  return [value, set, unset];
};

export default useSearchParam;
