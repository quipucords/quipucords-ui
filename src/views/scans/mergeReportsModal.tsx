import React, { useEffect, useRef } from 'react';
import { useTranslation } from 'react-i18next';
import { Bullseye, EmptyState, EmptyStateBody, Spinner } from '@patternfly/react-core';
import { Modal, ModalVariant } from '@patternfly/react-core/deprecated';
import { ExclamationCircleIcon, CheckCircleIcon } from '@patternfly/react-icons';
import { useMergeReportsApi } from '../../hooks/useScanApi';

interface MergeReportsModalProps {
  isOpen: boolean;
  reportIds: number[];
  onClose?: () => void;
  onSuccess?: (mergedReportId: number) => void;
}

const MergeReportsModal: React.FC<MergeReportsModalProps> = ({
  isOpen,
  reportIds,
  onClose = Function.prototype,
  onSuccess = Function.prototype
}) => {
  const { t } = useTranslation();
  const { requestReportsMerge, cancelReportsMerge, mergeProcessState } = useMergeReportsApi();
  // Workaround the following problem:
  // 1. onSuccess is called. This causes report to download and sets reportIdsToMerge to []
  // 2. empty reportIdsToMerge closes the modal
  // 3. that causes useEffect(..., [mergeProcessState]) to run BEFORE other useEffect manages to
  //    call cancelReportsMerge
  // 4. so onSuccess is called again, causing report to be downloaded again
  // The alternative is to have single useEffect depending on both isOpen and mergeProcessState,
  // and much more complex conditionals inside.
  const initiatedDownload = useRef<boolean>(false);

  useEffect(() => {
    if (isOpen && reportIds.length) {
      initiatedDownload.current = false;
      requestReportsMerge(reportIds);
    } else {
      cancelReportsMerge();
    }
  }, [isOpen, requestReportsMerge, cancelReportsMerge, reportIds]);

  useEffect(() => {
    if (mergeProcessState.state === 'Successful' && !initiatedDownload.current) {
      initiatedDownload.current = true;
      onSuccess(mergeProcessState.mergedReportId);
    }
  }, [mergeProcessState, onSuccess]);

  return (
    <Modal
      variant={ModalVariant.medium}
      title={t('merge.modal', { context: 'title' })}
      isOpen={isOpen}
      onClose={() => onClose()}
    >
      {mergeProcessState.state === 'InProgress' && (
        <Bullseye>
          <EmptyState
            headingLevel="h2"
            icon={Spinner}
            titleText={t('merge.modal', { context: 'body-in-progress' })}
          ></EmptyState>
        </Bullseye>
      )}
      {mergeProcessState.state === 'Successful' && (
        <Bullseye>
          <EmptyState
            headingLevel="h2"
            icon={CheckCircleIcon}
            titleText={t('merge.modal', { context: 'header-successful' })}
          >
            <EmptyStateBody>{t('merge.modal', { context: 'body-successful' })}</EmptyStateBody>
          </EmptyState>
        </Bullseye>
      )}
      {mergeProcessState.state === 'Errored' && (
        <Bullseye>
          <EmptyState
            headingLevel="h2"
            icon={ExclamationCircleIcon}
            titleText={t('merge.modal', { context: 'header-errored' })}
          >
            <EmptyStateBody>
              {t('merge.modal', {
                context: 'body-errored',
                msg: mergeProcessState.errorMessage
              })}
            </EmptyStateBody>
          </EmptyState>
        </Bullseye>
      )}
    </Modal>
  );
};

export { MergeReportsModal as default, MergeReportsModal, type MergeReportsModalProps };
