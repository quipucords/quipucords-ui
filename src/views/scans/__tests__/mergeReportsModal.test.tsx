import React from 'react';
import { render } from '@testing-library/react';
import { shallowComponent } from '../../../../config/jest.setupTests';
import * as useScanApi from '../../../hooks/useScanApi';
import { MergeReportsModal } from '../mergeReportsModal';

jest.mock('../../../hooks/useScanApi');

describe('MergeReportsModal', () => {
  let mockRequestReportsMerge;
  let mockCancelReportsMerge;

  const mockHookWithState = (mergeProcessState: useScanApi.MergeProcessState) => {
    (useScanApi.useMergeReportsApi as jest.Mock).mockReturnValue({
      requestReportsMerge: mockRequestReportsMerge,
      cancelReportsMerge: mockCancelReportsMerge,
      mergeProcessState: mergeProcessState
    });
  };

  beforeEach(() => {
    mockRequestReportsMerge = jest.fn();
    mockCancelReportsMerge = jest.fn();
    mockHookWithState({ state: 'InProgress' });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render a basic component', async () => {
    const props = {
      isOpen: true,
      reportIds: []
    };
    const component = await shallowComponent(<MergeReportsModal {...props} />);
    expect(component).toMatchSnapshot('basic');
  });

  it('should render a component after merge succeeded', async () => {
    mockHookWithState({ state: 'Successful', mergedReportId: 123 });
    const props = {
      isOpen: true,
      reportIds: []
    };
    const component = await shallowComponent(<MergeReportsModal {...props} />);
    expect(component).toMatchSnapshot('success');
  });

  it('should render a component after merge failed', async () => {
    mockHookWithState({ state: 'Errored', errorMessage: 'Lorem ipsum dolor sit' });
    const props = {
      isOpen: true,
      reportIds: []
    };
    const component = await shallowComponent(<MergeReportsModal {...props} />);
    expect(component).toMatchSnapshot('errored');
  });

  it('should request merge when modal is opened', async () => {
    const { rerender } = render(<MergeReportsModal isOpen={false} reportIds={[]} />);

    expect(mockRequestReportsMerge).not.toHaveBeenCalled();

    rerender(<MergeReportsModal isOpen={true} reportIds={[3, 2]} />);

    expect(mockRequestReportsMerge).toHaveBeenCalledTimes(1);
    expect(mockRequestReportsMerge).toHaveBeenCalledWith([3, 2]);
  });

  it('should cancel merge when modal is closed', async () => {
    const { rerender } = render(<MergeReportsModal isOpen={true} reportIds={[1]} />);

    expect(mockCancelReportsMerge).not.toHaveBeenCalled();

    rerender(<MergeReportsModal isOpen={false} reportIds={[]} />);

    expect(mockCancelReportsMerge).toHaveBeenCalledTimes(1);
  });

  it('should run onSuccess when merge finished', async () => {
    const mockOnSuccess = jest.fn();
    const mergedReportId = 123;
    const props = {
      isOpen: true,
      reportIds: [3, 2, 1],
      onSuccess: mockOnSuccess
    };
    const { rerender } = render(<MergeReportsModal {...props} />);
    mockHookWithState({ state: 'Successful', mergedReportId: mergedReportId });

    rerender(<MergeReportsModal {...props} />);

    expect(mockOnSuccess).toHaveBeenCalledTimes(1);
    expect(mockOnSuccess).toHaveBeenCalledWith(mergedReportId);
  });

  it('should run onSuccess only once while open', async () => {
    const mockOnSuccess = jest.fn();
    const mergedReportId = 123;
    const props = {
      isOpen: true,
      reportIds: [3, 2, 1],
      onSuccess: mockOnSuccess
    };
    const { rerender } = render(<MergeReportsModal {...props} />);
    mockHookWithState({ state: 'Successful', mergedReportId: mergedReportId });

    rerender(<MergeReportsModal {...props} />);
    // In view, isOpen is derived from reportIds. However, component will re-render once when reportIds is
    // reset to empty array, but before isOpen useEffect requests reset of hook state
    rerender(<MergeReportsModal {...props} reportIds={[]} />);

    expect(mockOnSuccess).toHaveBeenCalledTimes(1);
    expect(mockOnSuccess).toHaveBeenCalledWith(mergedReportId);
  });
});
