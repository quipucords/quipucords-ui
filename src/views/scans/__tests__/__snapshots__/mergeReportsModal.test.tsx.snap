// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MergeReportsModal should render a basic component: basic 1`] = `
<Modal
  actions={[]}
  appendTo={[Function]}
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className=""
  hasNoBodyWrapper={false}
  isOpen={true}
  onClose={[Function]}
  ouiaSafe={true}
  position="default"
  showClose={true}
  title="t([
  "merge.modal",
  {
    "context": "title"
  }
])"
  titleIconVariant={null}
  titleLabel=""
  variant="medium"
>
  <Bullseye>
    <EmptyState
      headingLevel="h2"
      icon={[Function]}
      titleText="t([
  "merge.modal",
  {
    "context": "body-in-progress"
  }
])"
    />
  </Bullseye>
</Modal>
`;

exports[`MergeReportsModal should render a component after merge failed: errored 1`] = `
<Modal
  actions={[]}
  appendTo={[Function]}
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className=""
  hasNoBodyWrapper={false}
  isOpen={true}
  onClose={[Function]}
  ouiaSafe={true}
  position="default"
  showClose={true}
  title="t([
  "merge.modal",
  {
    "context": "title"
  }
])"
  titleIconVariant={null}
  titleLabel=""
  variant="medium"
>
  <Bullseye>
    <EmptyState
      headingLevel="h2"
      icon={[Function]}
      titleText="t([
  "merge.modal",
  {
    "context": "header-errored"
  }
])"
    >
      <EmptyStateBody>
        t([
  "merge.modal",
  {
    "context": "body-errored",
    "msg": "Lorem ipsum dolor sit"
  }
])
      </EmptyStateBody>
    </EmptyState>
  </Bullseye>
</Modal>
`;

exports[`MergeReportsModal should render a component after merge succeeded: success 1`] = `
<Modal
  actions={[]}
  appendTo={[Function]}
  aria-describedby=""
  aria-label=""
  aria-labelledby=""
  className=""
  hasNoBodyWrapper={false}
  isOpen={true}
  onClose={[Function]}
  ouiaSafe={true}
  position="default"
  showClose={true}
  title="t([
  "merge.modal",
  {
    "context": "title"
  }
])"
  titleIconVariant={null}
  titleLabel=""
  variant="medium"
>
  <Bullseye>
    <EmptyState
      headingLevel="h2"
      icon={[Function]}
      titleText="t([
  "merge.modal",
  {
    "context": "header-successful"
  }
])"
    >
      <EmptyStateBody>
        t([
  "merge.modal",
  {
    "context": "body-successful"
  }
])
      </EmptyStateBody>
    </EmptyState>
  </Bullseye>
</Modal>
`;
